src/index.ts

```diff
@@ -586,40 +617,78 @@ export async function startAnalysis<
-                    if (targetOptions?.startEmulatorOptions?.headless === true) cmdArgs.push('-no-window');
-                    if (targetOptions?.startEmulatorOptions?.audio !== true) cmdArgs.push('-no-audio');
-                    if (targetOptions?.startEmulatorOptions?.ephemeral !== false) cmdArgs.push('-no-snapshot-save');
-                    if (targetOptions?.startEmulatorOptions?.hardwareAcceleration?.mode)
-                        cmdArgs.push('-accel', targetOptions?.startEmulatorOptions?.hardwareAcceleration?.mode);
-                    if (targetOptions?.startEmulatorOptions?.hardwareAcceleration?.gpuMode)
-                        cmdArgs.push('-gpu', targetOptions?.startEmulatorOptions?.hardwareAcceleration?.gpuMode);
-
-                    // eslint-disable-next-line require-atomic-updates
-                    emulatorProcess = (await startEmulator(emulatorName, cmdArgs))();
-                    emulatorProcess.catch((error) => {
-                        // We need to rethrow the error in this context to halt execution.
-                        throw new Error(`Emulator failed to start: ${error.message}`);
-                    });
-                    await platform.waitForDevice(150);
+
+                // ESLint wrongly thinks an optional chain would be logically equivalent
+                // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+                if (emulator) {
+                    // The emulator is not running anymore, but has already been created. We are restarting.
+                    if (emulator.failedStarts < (targetOptions?.startEmulatorOptions?.attemptRestarts ?? 1) + 1) {
+                        await emulator.start({ forceRestart: !!ensureOptions?.forceRestart });
+                    } else if (
+                        emulator.createdByLib &&
+                        emulator.rebuilds < (targetOptions?.createEmulator?.attemptRebuilds ?? 0)
+                    ) {
+                        await emulator.rebuild();
+                        await emulator.start();
+                    } else {
+                        throw new EmulatorError(`The emulator cannot be restarted after a failed run, because the restart and rebuild limits have been reached. Your emulator configuration is likely broken.
+Failed starts of the current build: ${emulator.failedStarts}
+Attempted rebuilds: ${emulator.rebuilds}
+Message of the last error: ${emulator.lastError?.message}`);
+                    }
+                } else if (!emulator) {
+                    if (targetOptions?.createEmulator !== undefined || targetOptions?.emulatorName) {
+                        // Create or start a new emulator
+                        // eslint-disable-next-line require-atomic-updates
+                        emulator = await AndroidEmulator.fromRunTarget(targetOptions);
+                        await emulator.start();
+                    }
                 }
```

Comment:
I think I would have put all that into `AndroidEmulator.start()`, but maybe I'm missing something?

Reply:
I guess this is a conceptual question? I felt like, if I keep AndroidEmulator free of "business logic", I can provide a robust implementation I can rely an build on.

Reply:
Exactly! In my view, `AndroidEmulator` should ensure to provide you with a working and started emulator matching the given parameters when you call `start()`. That of course includes restarting it if the first start fails. Your "business logic" can then be sure that it can call `adb` commands without having to care about the emulator state.